File: algorithmic_toolbox/week2_algorithmic_warmup/7_last_digit_of_the_sum_of_fibonacci_numbers_again/FibonacciPartialSum.java
import java.util.Scanner;

public class FibonacciPartialSum {
  private static long getFibonacciPartialSumNaive(long from, long to) {
    long sum = 0;

    long current = 0, next = 1;

    for (long i = 0; i <= to; ++i) {
      if (i >= from)
        sum += current;

      long new_current = next;
      next = next + current;
      current = new_current;
    }

    return sum % 10;
  }

  private static final int PISANO_PERIOD = 60;

  /*
   * The last digit of the numbers in the Fibonacci Sequence form a pattern that
   * repeats after
   * every 60th number. Source:
   * https://www.goldennumber.net/fibonacci-60-repeating-pattern/
   */
  private static long getFibonacciPartialSum(long from, long to) {
    long[] fibNums = new long[PISANO_PERIOD];
    fibNums[0] = 0;
    fibNums[1] = 1;

    for (int i = 2; i <= PISANO_PERIOD; i++)
      fibNums[i] = fibNums[i - 1] + fibNums[i - 2];

    from %= PISANO_PERIOD;
    to %= PISANO_PERIOD;

    long sum = 0;
    for (long j = from; j <= to; j++) {
      sum += fibNums[(int) j % PISANO_PERIOD];
      sum %= 10;
    }

    return sum;
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    long from = scanner.nextLong();
    long to = scanner.nextLong();
    scanner.close();

    System.out.println(getFibonacciPartialSum(from, to));
    System.out.println(getFibonacciPartialSumNaive(from, to));
  }
}

File: algorithmic_toolbox/week2_algorithmic_warmup/7_last_digit_of_the_sum_of_fibonacci_numbers_again/fibonacci_partial_sum.py
import sys
from typing import Final


def fibonacci_partial_sum_naive(from_: int, to: int) -> int:
    sum_ = 0

    current = 0
    next_ = 1

    for i in range(to + 1):
        if i >= from_:
            sum_ += current

        current, next_ = next_, current + next_

    return sum_ % 10


PISANO_PERIOD: Final = 60


def fibonacci_partial_sum(from_: int, to: int) -> int:
    """The last digit of the numbers in the Fibonacci Sequence form a pattern that repeats after every 60th number.

    Source: https://www.goldennumber.net/fibonacci-60-repeating-pattern/"""

    fib_nums_0_through_60: list[int] = [0, 1]

    for i in range(2, PISANO_PERIOD):
        fib_nums_0_through_60.append(fib_nums_0_through_60[i - 1] + fib_nums_0_through_60[i - 2])

    from_ %= PISANO_PERIOD
    to %= PISANO_PERIOD

    sum_ = 0
    for j in range(from_, to + 1):
        sum_ += fib_nums_0_through_60[j % PISANO_PERIOD]
        sum_ %= 10

    return sum_


if __name__ == "__main__":
    from_, to = map(int, sys.stdin.read().split())
    print(fibonacci_partial_sum(from_, to))
    print(fibonacci_partial_sum_naive(from_, to))
